File: algorithms/1130.last_stone_weight_ii/last_stone_weight_ii.cpp
class Solution {
public:
    int lastStoneWeightII(const vector<int>& stones) {
        const int sum = accumulate(stones.begin(), stones.end(), 0);
        const int target = (sum >> 1);
        int dp[3000] = {0};
        for (const int&stone: stones) {
            for (int j = target; j >= stone; --j) {
                dp[j] = max(dp[j], dp[j - stone] + stone);
            }
        }
        return sum - (dp[target] << 1);
    }
};

File: algorithms/1130.last_stone_weight_ii/last_stone_weight_ii.rs
impl Solution {
    pub fn last_stone_weight_ii(stones: Vec<i32>) -> i32 {
        let sum:i32 = stones.iter().sum();
        let target = sum as usize/2;
        let mut dp = vec![0; target as usize + 1];
        
        stones.iter().for_each(|&stone| {
            for j in (stone as usize..=target).rev() {
                dp[j] = dp[j].max(dp[j - stone as usize] + stone as usize);
            }
        });
        sum - dp[target] as i32 * 2
    }
}